### ISSUE TITLE
DuckDB goes into fatal mode when it fails to write parquet data in a mounted file system

### ISSUE THREAD

**Issue by osalloum** [duckdb/duckdb#12335]
### What happens?

When DuckDB fails to write a file to a mounted file system, it goes into fatal mode 

DuckDB is used as purely in memory database

Real world example, we are using mount-s3 ( https://github.com/awslabs/mountpoint-s3 ) to enable duckdb to read / write to an s3 bucket, this report duplicates and extends this issue https://github.com/duckdb/duckdb-java/issues/16 with a replicatable scenario

### To Reproduce

Create a mounted file system using fuse which throws a random error on fsync

This is an example of a passthrough mirroring using fuse 
```
#!/usr/bin/env python

from __future__ import with_statement

import errno
import os
import sys
from random import random

from fuse import FUSE, FuseOSError, Operations


class Passthrough(Operations):
    def __init__(self, root):
        self.root = root

    # Helpers
    # =======

    def _full_path(self, partial):
        partial = partial.lstrip("/")
        path = os.path.join(self.root, partial)
        return path

    # Filesystem methods
    # ==================

    def access(self, path, mode):
        #print("access %s %s", path, mode)
        full_path = self._full_path(path)
        if not os.access(full_path, mode):
            # print "access denied to %s" % full_path
            print("access denied to %s" % full_path)
            raise FuseOSError(errno.EACCES)

    def chmod(self, path, mode):
        full_path = self._full_path(path)
        return os.chmod(full_path, mode)

    def chown(self, path, uid, gid):
        full_path = self._full_path(path)
        return os.chown(full_path, uid, gid)

    def getattr(self, path, fh=None):
        full_path = self._full_path(path)
        st = os.lstat(full_path)
        return dict(
            (key, getattr(st, key))
            for key in (
                "st_atime",
                "st_ctime",
                "st_gid",
                "st_mode",
                "st_mtime",
                "st_nlink",
                "st_size",
                "st_uid",
            )
        )

    def readdir(self, path, fh):
        full_path = self._full_path(path)

        dirents = [".", ".."]
        if os.path.isdir(full_path):
            dirents.extend(os.listdir(full_path))
        for r in dirents:
            yield r

    def readlink(self, path):
        pathname = os.readlink(self._full_path(path))
        if pathname.startswith("/"):
            # Path name is absolute, sanitize it.
            return os.path.relpath(pathname, self.root)
        else:
            return pathname

    def mknod(self, path, mode, dev):
        return os.mknod(self._full_path(path), mode, dev)

    def rmdir(self, path):
        full_path = self._full_path(path)
        return os.rmdir(full_path)

    def mkdir(self, path, mode):
        return os.mkdir(self._full_path(path), mode)

    def statfs(self, path):
        full_path = self._full_path(path)
        stv = os.statvfs(full_path)
        return dict(
            (key, getattr(stv, key))
            for key in (
                "f_bavail",
                "f_bfree",
                "f_blocks",
                "f_bsize",
                "f_favail",
                "f_ffree",
                "f_files",
                "f_flag",
                "f_frsize",
                "f_namemax",
            )
        )

    def unlink(self, path):
        return os.unlink(self._full_path(path))

    def symlink(self, name, target):
        return os.symlink(name, self._full_path(target))

    def rename(self, old, new):
        return os.rename(self._full_path(old), self._full_path(new))

    def link(self, target, name):
        return os.link(self._full_path(target), self._full_path(name))

    def utimens(self, path, times=None):
        return os.utime(self._full_path(path), times)

    # File methods
    # ============

    def open(self, path, flags):
        full_path = self._full_path(path)
        return os.open(full_path, flags)

    def create(self, path, mode, fi=None):
        full_path = self._full_path(path)
        return os.open(full_path, os.O_WRONLY | os.O_CREAT, mode)

    def read(self, path, length, offset, fh):
        os.lseek(fh, offset, os.SEEK_SET)
        return os.read(fh, length)

    def write(self, path, buf, offset, fh):
        os.lseek(fh, offset, os.SEEK_SET)
        return os.write(fh, buf)

    def truncate(self, path, length, fh=None):
        full_path = self._full_path(path)
        with open(full_path, "r+") as f:
            f.truncate(length)

    def flush(self, path, fh):
        return os.fsync(fh)

    def release(self, path, fh):
        return os.close(fh)

    def fsync(self, path, fdatasync, fh):
        # fail random writes
        if fdatasync != 0 and random() < 0.5:
            return -errno.EIO
        return self.flush(path, fh)


def main(mountpoint, root):
    FUSE(Passthrough(root), mountpoint, nothreads=True, foreground=True)


if __name__ == "__main__":
    main(sys.argv[2], sys.argv[1])
```

create a directory which has the files
```
mkdir real_fs
```

Launch the mounted file system 
```
python3 real_fs mount
```

Launch duckdb CLI or via any language and start running queries until you get the Fatal error then try do any query

duckdb 
```
D copy (select * from generate_series(1, 1000) as t(a) cross join (select * from generate_series(1, 2) as t(b))  ) to 'mount/p2/' (FORMAT parquet, partition_by( b)) ;
IO Error: failed to create mount/p2/b=1/data_0.parquet, file exists! Enable OVERWRITE_OR_IGNORE option to force writing
D copy (select * from generate_series(1, 1000) as t(a) cross join (select * from generate_series(1, 2) as t(b))  ) to 'mount/p3/' (FORMAT parquet, partition_by( b)) ;
FATAL Error: fsync failed!
D select 1 as a;
FATAL Error: Failed: database has been invalidated because of a previous fatal error. The database must be restarted prior to being used again.
Original error: "fsync failed!"
D select 1 as a;
FATAL Error: Failed: database has been invalidated because of a previous fatal error. The database must be restarted prior to being used again.
Original error: "Failed: database has been invalidated because of a previous fatal error. The database must be restarted prior to being used again.
Original error: "fsync failed!""
```

### OS:

any

### DuckDB Version:

v0.10.4-dev124 cf5b770ccb

### DuckDB Client:

any

### Full Name:

Omar Ayoub Salloum

### Affiliation:

Holidu Gmbh

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a nightly build

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by carlopi** [comment]
I think the implementation of FileSync should probably throw IOException, to be possibly upgraded to FatalException by the caller when this is known to be problematic, this is done also by Windows implementation, so it should probably be OK.

**Comment by carlopi** [comment]
I think this solved by #12335, cool!
